### PR TITLE
Implement hash processor AIR constraints

### LINKED ISSUE #179: Finish implementing AIR constraints for the auxiliary table trace
AIR constraints need to be implemented for the following co-processors once #178 is complete:

- [x] write up the details of the modified transition constraint format discussed in the [Auxiliary Table Design note](https://hackmd.io/@grjte/ByKCo1Pr5) to ensure that:
    -  the changed format doesn't cause a problem with the constraints
    -  we know the tradeoff between the two sets of constraints in terms of number of constraints, degree of constraints, and number of multiplications.
- [x] hasher constraints
- [x] bitwise constraints
- [x] pow2 constraints

### PR DESCRIPTION
**Pull request by grjte**
This PR closes #179 , finishing all AIR constraints for the auxiliary table.

It contains the following changes:
- adds the original design docs for the hash processor
- updates the hash processor design with some required constraint changes
- implements the air constraints
- adds integration tests and basic unit test for the hash processor's AIR constraints

### CODE REVIEW

File: air/src/aux_table/hasher/mod.rs [hunk 1]
```diff
@@ -0,0 +1,616 @@
+use crate::utils::{are_equal, binary_not, is_binary, EvaluationResult};
+
+use super::{EvaluationFrame, Felt, FieldElement, HASHER_TRACE_OFFSET};
+use core::ops::Range;
+use vm_core::{
+    hasher::{
+        Hasher, CAPACITY_LEN, DIGEST_LEN, DIGEST_RANGE, NUM_ROUNDS, NUM_SELECTORS, STATE_WIDTH,
+    },
+    utils::range as create_range,
```
Comment (bobbinth) [inline air/src/aux_table/hasher/mod.rs:9]: Thought for the future: I wonder if we should create something like `RangeExt` trait and implement something like `new()` function on it. Might be easier than using this `range()` function.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks great! Thank you! I left a couple of very small comments inline.

File: air/src/lib.rs [hunk 2]
```diff
@@ -137,7 +130,7 @@ impl Air for ProcessorAir {
     // TRANSITION CONSTRAINTS
     // --------------------------------------------------------------------------------------------
 
-    fn evaluate_transition<E: FieldElement<BaseField = Felt>>(
+    fn evaluate_transition<E: FieldElement<BaseField = Felt> + From<Felt>>(
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/lib.rs:133]: Yeah, not sure why this ended up getting introduced - it's not necessary, and I've removed it.

File: air/src/aux_table/mod.rs [hunk 3]
```diff
@@ -56,24 +67,35 @@ pub fn get_transition_constraint_degrees() -> Vec<TransitionConstraintDegree> {
...
 pub fn get_transition_constraint_count() -> usize {
     NUM_CONSTRAINTS
+        + hasher::get_transition_constraint_count()
         + bitwise_pow2::get_transition_constraint_count()
         + memory::get_transition_constraint_count()
 }
 
 /// Enforces constraints for the auxiliary table and all of its co-processors.
-pub fn enforce_constraints<E: FieldElement>(
+pub fn enforce_constraints<E: FieldElement + From<Felt>>(
```
[the remark this answers is not in the document]
Comment (grjte) [reply air/src/aux_table/mod.rs:76]: Yep, agreed. The examples in Winterfell use `From`, which is why I had done it that way in the hasher, but the consistency of `<BaseField = Felt>` is better. I've updated it

File: air/src/aux_table/hasher/mod.rs (hunk 1 above)
Comment (grjte) [reply air/src/aux_table/hasher/mod.rs:9]: Yeah, I think this makes sense to introduce in a future PR

**Review by bobbinth — APPROVED** [review]
Everything looks great! Thank you!

**Pull request merged**